### PR TITLE
Add pthread for linux

### PR DESCRIPTION
**Pull request by fxzjshm**
Maybe this time it is a better solution.
Verified on Ubuntu 16.04, macOS (both are in virtual machine) and Bash on Windows.
Sorry for disturbing again.

### CODE REVIEW

**Comment by fxzjshm** [comment]
@intrigus could you review, please?

**Comment by intrigus** [comment]
Did you have a look at the `-pthread` cli argument? 
This should also add some useful defines, which should make the code re-entrant safe. 
Sorry for the delay.

**Comment by fxzjshm** [comment]
It seems that [in this file](https://github.com/jtransc/jtransc/blob/master/jtransc-gen-cpp/src/com/jtransc/gen/cpp/CppCompiler.kt#L34) `-pthread` is not used with cmake.

> This should also add some useful defines, which should make the code re-entrant safe.

I don't really understand this, perhaps because I seldom use thread. Could you please say a bit clearer?

**Comment by intrigus** [comment]
The cmake integration has been added after the other compilers.
That's why it's probably not using `-pthread`.

>I don't really understand this, perhaps because I seldom use thread. Could you please say a bit clearer?
It's should make the standard c library thread safe so you can call certain functions from multiple threads. E.g. opening a file. Another example is the `errno` variable which contains a code when an error happens. When `-pthread` is specified every `errno` is actually a variable per thread, so every thread has its own `errno` variable.

Also have a look at this question https://stackoverflow.com/questions/41267121/pthread-lpthread-and-minimal-dynamic-linktime-dependencies __and both answers__

**Comment by fxzjshm** [comment]
Hi @intrigus , 
I've implemened MonitorEnter and MonitorExit using recursive_mutex of C++. 
I also added some 'synchronized' into some methods. 
Could you please tell me **whether it is far from thread-safe**?

**Comment by soywiz** [comment]
Great! Also finally test passes. I will test on mac this afternoon. Since we know it works on Linux and Windows at least.

I think (dont remember, so not fully sure until I recheck) that synchronized methods doesn't explicitly use MonitorEnter/MonitorExit but they use a flag in the method info to specify they are synchronized. We have two options here: detect that flag in each target and do the proper, or try to convert the ast to call those methods. The problem is that with returns in the middle it would be more complicated or would require extra methods. So I would propose at least initially to implement in each target.
I will verify this and will help implementing it if it is required

**Comment by fxzjshm** [comment]
Sorry I didn't realized this problem. 
Now I have 2 ideas:

1. Use statements like:
```C
_ReturnType func_Wrapper(_ArgType[] args){
    monitorEnter();
    _ReturnType ret = func(args);
    monitorExit();
    return ret;
}
```
We can control this when generating either ast or actual code. But I know little about both...

2. Use 
```Java
monitorEnter();
try{
    // other operations of this method
    // It seems that in Java, return also goes to finally. 
    // But other languages? I'm sorry I don't know.
}finally{
    monitorExit();
}
```
But I don't know if every target supports syntax like this.

**Comment by soywiz** [comment]
Thats right. That is what i was going to propose. 
Older versions of PHP for example didnt have finally. Newer versions do have. But PHP for example doesnt support threads in the first place. So it is not a problem. I think that most or all the targets supporting threads will have try finally.

You can check C# code generator and search for `unchecked` as far as i remember it wraps the whole function. So you can do the same in C++ to wrap the whole function with try finally for methods that have the synchronized flag

**Comment by soywiz** [comment]
Method has a modifiers property. And you fan check if the method is synchronized like this:
https://github.com/jtransc/jtransc/blob/master/jtransc-core/src/com/jtransc/ast/ast.kt#L883

**Comment by fxzjshm** [comment]
Well life with Kotlin is quite easier 😄
And your code is much more simple than mine.
I'll try.

**Comment by fxzjshm** [comment]
Is this far from enough to support sync?

**Comment by soywiz** [comment]
Cool! Looks good to me. I will rethink the haxe part. Didnt thought about super. But at least c++ and c# targets already support it!

File: jtransc-gen-d/src/com/jtransc/gen/d/DTarget.kt [hunk 1]
```diff
@@ -355,4 +355,17 @@ class DGenerator(injector: Injector) : CommonGenerator(injector) {
 	override fun genExprCastChecked(e: String, from: AstType.Reference, to: AstType.Reference): String {
 		return "checkCast!(${to.targetNameRef})($e)"
 	}
+
+	override fun genBody2WithFeatures(method: AstMethod, body: AstBody): Indenter = Indenter {
+		if(method.modifiers.isSynchronized) {
+			line("try{")
```
Comment (intrigus) [inline jtransc-gen-d/src/com/jtransc/gen/d/DTarget.kt]: This is almost the same in all targets why has this not been de-duplicatd?

Comment (fxzjshm) [inline jtransc-gen-d/src/com/jtransc/gen/d/DTarget.kt]: Though many targets have `try{ ... } finally { ... }` syntax, not all targets support it. So I am afraid I cannot put it into CommonGenerator.

**Comment by soywiz** [comment]
I merged changing the title. Though i should have renamed the PR too. Initially i just approved the PR, but noticed that fxjshm cant merge it. I invited you to the org sometime ago. Did you receive the invitation?

**Comment by fxzjshm** [comment]
I saw it. I'm afraid I'm just too young, I'm busy with my schoolwork so that I cannot contribute to this project regularly these years. Maybe I will come back after 3 years (when I get into a university). Thank you, **all of you**, very much.

**Comment by soywiz** [comment]
No problem! And thanks for all your contributions so far :) Of course any other contribution is always welcome. Being part of the org was just to give you the right credit for your contributions (that even if are just infew in a specific timeframe have their own value), to allow you to merge your stuff or reviewing other things only if you want, and afaik it doesn't send you extra notifications if you do not watch the repo. Didnt want to make any pressure at all! You can disappear and leave at will at any time and join at any moment too for a specific contributions or not return at all; for me this is how opensource works :) The same goes for me; I have other projects too. Im in orgs i do not collaborate anymore because i lost interest or do not have enough time. Though in my case sice i created jtransc if there are critial issues about jtransc and nobody fixes then i will try to fix them asap even if in that period of time i have other priorities.

**Pull request merged**